src/mon/ConnectionTracker.cc

```diff
@@ -219,22 +216,27 @@ void ConnectionTracker::notify_rank_removed(int rank_removed)
     my_reports.current.erase(ci++);
     my_reports.history.erase(hi++);
   }
-  ceph_assert((my_reports.current.size() == starting_size) ||
-	      (my_reports.current.size() + 1 == starting_size));
+  ceph_assert((my_reports.current.size() == starting_size_current) ||
+    (my_reports.current.size() + 1 == starting_size_current));
 
-  // now move ranks down one in peer_reports
-  starting_size = peer_reports.size();
+  size_t starting_size = peer_reports.size();
   auto pi = peer_reports.upper_bound(rank_removed);
+  // Remove the target rank and adjust everything that comes after.
+  // Note that we don't adjust current and history for our peer_reports
+  // because it is better to rely on our peers on that information.
+  peer_reports.erase(rank_removed);
   while (pi != peer_reports.end()) {
-    peer_reports[pi->first - 1] = pi->second;
-    peer_reports.erase(pi++);
+    peer_reports[pi->first - 1] = pi->second; // copy content of next rank to ourself.
+    --peer_reports[pi->first - 1].rank; // update the rank of our new content.
```

Comment:
If we don't also update the rank of our peer locally after we move our peers, then it could be a problem so that's why I'm also updating the rank

Reply:
Can you explain this "could be a problem"?
I'm not sure futzing with these ranks is an issue, but I really want them to be read-only. I did some reasonably solid proofs on convergence with gossip-based updates to these but it was all based on treating the peer_reports as a vector clock thing.

Reply:
if we don't want to change things in peer_reports, then we might have to change how we do things in
ConnectionTracker::reports or ConnectionTracker::receive_peer_report. What we might want to do is instead of passing `report.rank`, we can pass `i.first` (the key of peer_reports map)

Reply:
Sounds good, Addressed in new push